sklearn/cluster/k_means_.py

```diff
@@ -299,23 +299,22 @@ def k_means(X, n_clusters, init='k-means++', precompute_distances='auto',
                          precompute_distances)
 
     # subtract of mean of x for more accurate distance computations
-    if not sp.issparse(X) or hasattr(init, '__array__'):
-        X_mean = X.mean(axis=0)
     if not sp.issparse(X):
+        X_mean = X.mean(axis=0)
         # The copy was already done above
         X -= X_mean
 
-    if hasattr(init, '__array__'):
-        init = check_array(init, dtype=X.dtype.type, copy=True)
-        _validate_center_shape(X, n_clusters, init)
+        if hasattr(init, '__array__'):
+            init = check_array(init, dtype=X.dtype.type, copy=True)
+            _validate_center_shape(X, n_clusters, init)
 
-        init -= X_mean
-        if n_init != 1:
-            warnings.warn(
-                'Explicit initial center position passed: '
-                'performing only one init in k-means instead of n_init=%d'
-                % n_init, RuntimeWarning, stacklevel=2)
-            n_init = 1
+            init -= X_mean
```

Comment:
Sorry, could you clarify this again? Are you saying that it doesn't need to check if it's array in order to subtract the mean?